scipy/sparse/linalg/propack.py

```diff
@@ -0,0 +1,336 @@
+        lwork = m + n + 9*kmax + 2*kmax*kmax + 4 + max(m + n, 4*kmax + 4)
+        liwork = 2*kmax + 1
+    work = np.empty(lwork, dtype=typ.lower())
+    iwork = np.empty(liwork, dtype=np.int32)
+
+    # dummy arguments: these are passed to aprod, and not used in this wrapper
+    dparm = np.empty(1, dtype=typ.lower())
+    iparm = np.empty(1, dtype=np.int32)
+
+    if typ.isupper():
+        # PROPACK documentation is unclear on the required length of zwork.
+        # Use the same length Julia's wrapper uses
+        # see https://github.com/JuliaSmoothOptimizers/PROPACK.jl/
+        zwork = np.empty(m + n + 32*m, dtype=typ)
+
+        if irl_mode:
+            u, sigma, bnd, v, info = lansvd_irl(_which_converter[which],
+                                                jobu, jobv, m, n,
+                                                shifts, k, maxiter, aprod,
+                                                u, v, tol, work, zwork,
+                                                iwork, doption, ioption,
+                                                dparm, iparm)
+        else:
+            u, sigma, bnd, v, info = lansvd(jobu, jobv, m, n, k, aprod, u, v,
+                                            tol, work, zwork, iwork, doption,
+                                            ioption, dparm, iparm)
+    else:
+        if irl_mode:
+            u, sigma, bnd, v, info = lansvd_irl(_which_converter[which],
+                                                jobu, jobv, m, n,
+                                                shifts, k, maxiter, aprod,
+                                                u, v, tol, work, iwork,
+                                                doption, ioption, dparm,
+                                                iparm)
+        else:
+            u, sigma, bnd, v, info = lansvd(jobu, jobv, m, n, k, aprod, u, v,
+                                            tol, work, iwork, doption,
+                                            ioption, dparm, iparm)
+
+    if info > 0:
+        raise LinAlgError(
+            f"An invariant subspace of dimension {info} was found.")
+    elif info < 0:
+        raise LinAlgError(
+            f"k={k} singular triplets did not converge within "
+            f"kmax={kmax} iterations")
+
+    # info == 0: The K largest (or smallest) singular triplets were computed
+    # succesfully!
+
+    # catch corner case where A is all zeros and v is not orthogonal
+    if not np.any(v):
+        v = np.eye(*v.shape)
+
+    # construct return tuple based on caller's desired output
+    return {
+        (0, 0, 0): sigma,
+        (1, 0, 0): (u[:, :k], sigma),
+        (0, 1, 0): (sigma, v[:, :k].conj().T),
+        (1, 1, 0): (u[:, :k], sigma, v[:, :k].conj().T),
+        (0, 0, 1): (sigma, bnd),
+        (1, 1, 1): (u[:, :k], sigma, v[:, :k].conj().T, bnd),
+        (0, 1, 1): (sigma, v[:, :k].conj().T, bnd),
+        (1, 0, 1): (u[:, :k], sigma, bnd),
+    }[(compute_u, compute_v, full_output)]
```

Comment:
There is precedent for this sort of thing, but what do you think about a `namedtuple` that has `None` for entries that aren't computed?

Reply:
I think we're thinking we'll just expose PROPACK through `svds`, are you still leaning towards making this a `namedtuple`?

Reply:
It doesn't have to be a namedtuple here, but it should return `None` for entries that aren't computed. That's what `svds` needs, ultimately, so it's a lot simpler to just do it here then write logic that deals with different numbers of outputs.

Reply:
Never mind. There were a few problems here.

First, the interface is unusual by today's standards. With `return_singular_vectors=False`, nothing is returned for the singular vectors, whereas with `return_singular_vectors='u'` or `return_singular_vectors='vh'`, `None` _might_ be returned in their place. (I say _might_ because of point three, below.) Even if we consider an argument controlling the number of outputs to be OK, it is strange to me that sometimes `None` is returned and somethings _nothing_ is returned.

Second, this behavior wasn't documented. So when I wrote that comment, I thought that it was always returning `None` if the singular vectors were not calculated.

There is a third strange thing that _is_ (at least partially) documented: string values for this argument may be ignored depending on the shape of the matrix. So, for instance, if the matrix is 6x6 and we ask for `return_singular_vectors='vh'`, it will still calculate and return both the left and right singular vectors.

So I don't know what to do here except implement these rules very carefully.